File: src/NStreamCom/BufferedDecoder.cs
namespace NStreamCom
{
    public class BufferedDecoder
    {
        private static readonly int BUFFER_CLEAR_SIZE = 512;

        public int DecodedIndex { get; private set; } = 0;

        public byte[] Bytes { get => [.. _bytes]; }

        private readonly List<byte> _bytes = [];

        private byte _rightShift = 0;

        public void Reset()
        {
            DecodedIndex = 0;
            _rightShift = 0;
            if (BUFFER_CLEAR_SIZE <= _bytes.Count)
                _bytes.Clear();
        }

        public void Next(byte b, bool isLast)
        {
            if (!isLast)
                EnsureSize()[DecodedIndex] = (byte)(b >> _rightShift);

            if (_rightShift != 0)
                _bytes[DecodedIndex - 1] |= (byte)(b << (8 -  _rightShift));

            if (_rightShift == 7)
            {
                _rightShift = 0;
            }
            else
            {
                _rightShift++;
                DecodedIndex++;
            }
        }

        private List<byte> EnsureSize()
        {
            while (_bytes.Count <= DecodedIndex)
                _bytes.Add(0);
            return _bytes;
        }
    }
}

File: src/nstreamcom/buffered_decoder.py
from .bits import as_byte


class BufferedDecoder:
    def __init__(self):
        self._decoded_index: int = 0
        self._buffer: bytearray = bytearray()
        self._right_shift: int = 0

    def reset(self) -> None:
        self._decoded_index = 0
        self._right_shift = 0
        self._buffer.clear()

    @property
    def decoded_index(self) -> int:
        return self._decoded_index

    @property
    def bytearray(self) -> bytearray:
        return self._buffer

    def next(self, byte: int, is_last: bool) -> None:
        if not is_last:
            self._ensure_size[self._decoded_index] = as_byte(byte >> self._right_shift)

        if self._right_shift != 0:
            self._buffer[self._decoded_index - 1] |= as_byte(byte << (8 - self._right_shift))

        if self._right_shift == 7:
            self._right_shift = 0
        else:
            self._right_shift += 1
            self._decoded_index += 1

    @property
    def _ensure_size(self) -> bytearray:
        while len(self._buffer) <= self._decoded_index:
            self._buffer.append(0)
        return self._buffer
